### PR TITLE
[ENH] Issue 3345: Adding FreeSurfer longitudinal interfaces

### PR DESCRIPTION
**Pull request by l-espana**
## Summary

Adds FreeSurfer longitudinal interfaces per issue https://github.com/nipy/nipype/issues/3345.

## List of changes proposed in this PR (pull-request)

- Add BaseReconAll to allow running recon-all with timepoints to create a base template.
- Add LongReconAll to allow running recon-all for a session given a base template.

### CODE REVIEW

File: nipype/interfaces/freesurfer/longitudinal.py [hunk 1]
```diff
@@ -251,3 +253,200 @@ def _list_outputs(self):
...
+class LongReconAll(CommandLine):
+    """Uses FreeSurfer's longitudinal recon-all to process a subject given
+    the previously processed base template.
+
+    Examples
+    ---------
+
+    >>> from nipype.interfaces.freesurfer.longitudinal import LongReconAll
+    >>> longrecon = LongReconAll()
+    >>> longrecon.inputs.long_id = ("ses-1","sub-template")
```
Comment (ghisvail) [inline nipype/interfaces/freesurfer/longitudinal.py:379]: From a workflow construction perspective, providing `long_id` as tuple might make composition from `all`, `base` and then `long` interfaces more difficult.

In [pydra-freesurfer](https://github.com/aramis-lab/pydra-freesurfer/blob/main/pydra/tasks/freesurfer/recon_all.py), I have split it as `longitudinal_timepoint_id` and `longitudinal_template_id` and required the latter be specified if the former is. This way they are both exposed as separate input to the interface.

**Review by ghisvail — COMMENTED** [review]
This looks very good. I like your proposal for separating the base and long interfaces from the main one, which avoids overcharging the main interface which is already quite complex.

I was wondering whether you had tested a complete  `Nipype` workflow for FS longitudinal composed of all three tasks? I have made a suggestion which could improve composition.

File: nipype/interfaces/freesurfer/longitudinal.py (hunk 1 above)
Comment (l-espana) [reply nipype/interfaces/freesurfer/longitudinal.py:379]: Oh, that's a very good point. I can separate those out. I tend to run each step separately, but I will make that change and test out a full workflow with all three steps when I can...should definitely make sure that works with this setup.

**Comment by ghisvail** [comment]
> Update on this: I was able to successfully run a nipype workflow with all longitudinal steps.

That's great.

> Not sure if I should document the relevant pieces of that somewhere

One suggestion would be to post a working snippet here with like all 3 instances of `ReconAll`, `BaseReconAll` and `LongReconAll` wired together?

**Comment by l-espana** [comment]
This is hopefully enough to help people get started and should be close enough to what I actually ran...
```from nipype.pipeline import engine as pe
from nipype.interfaces.freesurfer.preprocess import ReconAll
from nipype.interfaces.freesurfer.longitudinal import BaseReconAll, LongReconall
from niworkflows.engine.workflows import LiterateWorkflow as Workflow
from niworkflows.utils.bids import collect_data

bids_dir = "."
subject_data = collect_data(bids_dir, "sub-1")[0]
freesurfer_dir = "derivatives/freesurfer"
wf = Workflow(name="longitudinal_freesurfer")

initrecon = pe.MapNode(interface=ReconAll(directive="all", subjects_dir=freesurfer_dir), 
    name="initrecon", iterfield=["T1_files", "subject_id"])
initrecon.inputs.subject_id = ["ses-1","ses-2"]
initrecon.inputs.T1_files = subject_data["t1w"]

baserecon = pe.Node(BaseReconAll(base_id="sub-1", directive="all", subjects_dir=freesurfer_dir), 
    name="baserecon")

longrecon = pe.MapNode(LongReconAll(directive="all", subjects_dir=freesurfer_dir), 
    name="longrecon", iterfield=["long_id"])
longrecon.inputs.long_id = ["ses-1","ses-2"]

wf.connect([
    (initrecon, baserecon, [("subject_id", "timepoints")]),
    (baserecon, longrecon, [("subject_id", "base_id")]),
])
```

File: .zenodo.json [hunk 2]
```diff
@@ -890,6 +890,11 @@
       "affiliation": "MIT, HMS",
       "name": "Ghosh, Satrajit",
       "orcid": "0000-0002-5312-6729"
+    },
+    {
+      "affiliation": "Department of Neurosurgery, Medical College of Wisconsin",
+      "name": "Espana, Lezlie",
+      "orcid": "0000-0002-6466-4653"
```
Comment (ghisvail) [inline .zenodo.json:897]: I'd suggest to modify the zenodo file within its own commit. Also, I believe you should keep Satra as last author.

Comment (l-espana) [reply .zenodo.json:897]: Oh of course, my mistake.

Comment (ghisvail) [reply .zenodo.json:897]: No problem

File: nipype/interfaces/freesurfer/longitudinal.py [hunk 3]
```diff
@@ -251,3 +253,207 @@ def _list_outputs(self):
         outputs = self.output_spec().get()
         outputs["out_file"] = os.path.abspath(self.inputs.out_file)
         return outputs
+
+
+class BaseReconAllInputSpec(ReconAllInputSpec):
+    subject_id = traits.Str(argstr="-subjid %s", desc="subject name")
```
Comment (ghisvail) [inline nipype/interfaces/freesurfer/longitudinal.py:259]: My understanding is that `subject_id` should not be specified when calling `recon-all -base` and it's the base template ID which supersedes it.

```suggestion
```

File: nipype/interfaces/freesurfer/longitudinal.py [hunk 4]
```diff
@@ -251,3 +253,207 @@ def _list_outputs(self):
...
+    base_id = traits.Str(argstr="-base %s", desc="base template name", xor=["subject_id"])
+    timepoints = InputMultiObject(
+        traits.Str(), argstr="-tp %s...", desc="processed time point to use in template"
+    )
```
Comment (ghisvail) [inline nipype/interfaces/freesurfer/longitudinal.py:263]: A few suggestions here

Comment (ghisvail) [reply nipype/interfaces/freesurfer/longitudinal.py:263]: This proposal adds more explicitness to the argument names as to what they refer to. Both points to subject identifiers in FreeSurfer's subjects directory, hence the `id` suffix. We also try to favour long form of arguments, hence `tp` -> `-base-tp`.

```suggestion
    base_template_id = traits.Str(
        argstr="-base %s",
        desc="base template name",
        mandatory=True,
    )
    base_timepoint_ids = InputMultiObject(
        traits.Str(),
        argstr="-base-tp %s...",
        desc="processed time point to use in template",
        mandatory=True,
    )
```

File: nipype/interfaces/freesurfer/longitudinal.py [hunk 5]
```diff
@@ -251,3 +253,207 @@ def _list_outputs(self):
...
+        directive = self.inputs.directive
+        if not isdefined(directive):
+            steps = []
+
+        iflogger.info(f"recon-all: {cmd}")
+        return cmd
+
+
+class LongReconAllInputSpec(ReconAllInputSpec):
+    subject_id = traits.Str(argstr="-subjid %s", desc="subject name")
```
Comment (ghisvail) [inline nipype/interfaces/freesurfer/longitudinal.py:355]: Same comment as for `BaseReconAll`, `subject_id` should not be exposed I believe.

```suggestion
```

File: nipype/interfaces/freesurfer/longitudinal.py [hunk 6]
```diff
@@ -251,3 +253,207 @@ def _list_outputs(self):
...
+        xor=["subject_id"],
+        requires=["base_id"],
+        position=1
+    )
+    base_id = traits.Str(
+        argstr="%s",
+        desc="longitudinal base template id",
+        requires=["long_id"],
+        position=2
+    )
```
Comment (ghisvail) [inline nipype/interfaces/freesurfer/longitudinal.py:368]: Similar comment as for `BaseReconAll` regarding namings and constraints.

```suggestion
    longitudinal_timepoint_id = traits.Str(
        argstr="-long %s",
        desc="longitudinal session/timepoint id",
        mandatory=True
        position=1
    )
    longitudinal_template_id = traits.Str(
        argstr="%s",
        desc="longitudinal base template id",
        mandatory=True,
        position=2
    )
```

File: nipype/interfaces/freesurfer/longitudinal.py [hunk 7]
```diff
@@ -251,3 +253,207 @@ def _list_outputs(self):
...
+                subject_id=subject_id, subjects_dir=subjects_dir, hemi=hemi
+            )._list_outputs()
+        )
+        outputs["subject_id"] = subject_id
+        outputs["subjects_dir"] = subjects_dir
+        return outputs
+
+    def _is_resuming(self):
+        subjects_dir = self.inputs.subjects_dir
+        subject_id = f"{self.inputs.long_id}.long{self.inputs.base_id}"
```
Comment (ghisvail) [inline nipype/interfaces/freesurfer/longitudinal.py:433]: Missing dot in computed `subject_id`:

```suggestion
        subject_id = f"{self.inputs.long_id}.long.{self.inputs.base_id}"
```

**Review by ghisvail — CHANGES_REQUESTED** [review]
Thanks for this initial proposal. This first iteration looks very good and your snippet shows very well how it can be all composed.

I have a few suggestions wrt to namings, constraints and some typos.

**Comment by ghisvail** [comment]
My suggestions did not include updates to the doctests, so they will likely fail. Could you update them please?

You will also need to refresg the test suite generated automatically from the specs.

**Comment by l-espana** [comment]
Yes, working on that now. :)

**Comment by l-espana** [comment]
I don't know that we can take out the subject_id from the input spec...We're inheriting from the ReconAll spec which has a default subject_id of "recon-all" so I think we need to override it. Unless there's another way to negate that?

**Comment by ghisvail** [comment]
> @[l-espana](https://github.com/l-espana) l-espana requested a review from [ghisvail](https://github.com/ghisvail)

I will have a look at it next week.

> I don't know that we can take out the subject_id from the input spec

If we can't, then that may be a sign Base and Long specs should be merged into the main ReconAll with some appropriate `xor` logic. That's what I did for the Pydra interface and it worked.

**Comment by effigies** [comment]
Please merge master to fix the tests.

**Comment by 0rC0** [comment]
> I was able to successfully run a nipype workflow with all longitudinal steps.
great! :-) 

>  not sure if I need to update anything else or if anybody else has comments, perhaps @0rC0? I know you had worked on this previously as well...

It looks very good to me. At the beginning of the new year I can do some more tests.

**Comment by l-espana** [comment]
Hi! Any chance we can get this reviewed?

**Comment by effigies** [comment]
@ghisvail Were you reviewing this? Or would you like me to?

File: nipype/interfaces/freesurfer/preprocess.py [hunk 8]
```diff
@@ -927,6 +930,29 @@ class ReconAllInputSpec(CommandLineInputSpec):
     )
     flags = InputMultiPath(traits.Str, argstr="%s", desc="additional parameters")
 
+    # Longitudinal runs
```
Comment (ghisvail) [inline nipype/interfaces/freesurfer/preprocess.py:933]: We might need to introduce additional constraints to avoid using cross-sectional specific arguments with base and long modes.

I am thinking of the `T1_files`, `T2_file` and `FLAIR_file` parameters at least. These should probably get a requirement on `subject_id`, which would achieve what we want thanks to your `xor` constraints.

**Review by ghisvail — CHANGES_REQUESTED** [review]
One last suggestion to improve the robustness of the interface.

The rest looks good to me. Nice work 👏

File: nipype/interfaces/freesurfer/preprocess.py (hunk 8 above)
Comment (l-espana) [reply nipype/interfaces/freesurfer/preprocess.py:933]: That didn't cross my mind but excellent thought. I've add some requires and will test some variations next week to make sure that works as expected.

Comment (ghisvail) [reply nipype/interfaces/freesurfer/preprocess.py:945]: Looks good to me :+1:

File: nipype/interfaces/freesurfer/preprocess.py [hunk 9]
```diff
@@ -842,21 +846,28 @@ class ReconAllInputSpec(CommandLineInputSpec):
...
         min_ver="5.3.0",
         desc="Convert T2 image to orig directory",
+        requires=["subject_id"],
     )
     FLAIR_file = File(
         exists=True,
         argstr="-FLAIR %s",
         min_ver="5.3.0",
         desc="Convert FLAIR image to orig directory",
+        requires=["subject_id"]
```
Comment (ghisvail) [inline nipype/interfaces/freesurfer/preprocess.py:870]: formatting

```suggestion
        requires=["subject_id"],
```

File: nipype/interfaces/freesurfer/preprocess.py [hunk 10]
```diff
@@ -885,18 +896,21 @@ class ReconAllInputSpec(CommandLineInputSpec):
...
+        requires=["subject_id"],
     )
     big_ventricles = traits.Bool(
         argstr="-bigventricles",
         desc=("For use in subjects with enlarged " "ventricles"),
     )
     brainstem = traits.Bool(
-        argstr="-brainstem-structures", desc="Segment brainstem structures"
+        argstr="-brainstem-structures", desc="Segment brainstem structures",
+        requires=["subject_id"],
```
Comment (ghisvail) [inline nipype/interfaces/freesurfer/preprocess.py:907]: formatting

```suggestion
        argstr="-brainstem-structures",
        desc="Segment brainstem structures",
        requires=["subject_id"],
```

**Review by ghisvail — APPROVED** [review]
Looks good to be merged. I have highlighted a few formatting nuggets which `black` might complain about.

**Comment by ghisvail** [comment]
@effigies I don't have the privileges to merge this PR. Could you take care of it? :pray:

**Comment by ghisvail** [comment]
> Looks like CI is broken; apologies. @ghisvail If you have time to look into this, it would be appreciated. Otherwise I'll get to it this week.

It looks like issues with CI started to surface on the last two merged PRs. CI complains that jobs no longer find `black` or `pytest` although they appear in the install step in the logs. The only reason I could think of is that the venv is no longer visible or activated between calls to the CI scripts (perhaps as a result of a change in GitHub Actions behaviour)?

Perhaps we could refactor CI to not use a dedicated venv? Maybe there are good reasons for it, I have just never seen it done in practice with GHA since each job is executed in a separate throwaway container anyway. It might also make things considerably simpler, as I find the current flow of CI scripts quite hard to follow. Let me know what you think.

**Comment by effigies** [comment]
Been a bit, but I got CI working (except CircleCI...) so hopefully any red checks you see are actionable.
